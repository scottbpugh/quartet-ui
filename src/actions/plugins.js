// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import {createAction} from "redux-actions";

export default {
  addToTreeServers: createAction("PLUGINS_ADD_TO_TREE_SERVERS"),
  removeFromTreeServers: createAction("PLUGINS_REMOVE_FROM_TREE_SERVERS"),
  addButtonToControls: createAction("PLUGINS_ADD_BUTTON_TO_CONTROLS"),
  removeButtonFromControls: createAction("PLUGINS_REMOVE_BUTTON_FROM_CONTROLS"),
  pluginEnabled: createAction("PLUGINS_NEW_PLUGIN_ENABLED"),
  pluginDisabled: createAction("PLUGINS_PLUGIN_DISABLED"),
  receivedPluginsData: createAction("PLUGINS_RECEIVED_PLUGINS_DATA"),
  addLocalPlugin: createAction("PLUGINS_ADD_LOCAL_PLUGIN"),
  pluginListUpdated: createAction("PLUGINS_PLUGIN_LIST_UPDATED"),
  resetPlugins: createAction("PLUGINS_RESET_PLUGINS")
};
window.qu4rtet.exports("actions/plugins", this);
