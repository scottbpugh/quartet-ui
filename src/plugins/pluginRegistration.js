// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
import React from "react";
import {Route} from "react-router";
import {injectIntl} from "react-intl";

/**
 * PluginRegistry - Holds references to various plugin objects
 * such as routes, reducers, components, and localization messages.
 * Also emits an event when a reducer is added.
 */
class PluginRegistry {
  constructor() {
    this._emitChange = null;

    /*
    Holds references to reducers that come from plugins.
    */
    this._reducers = {};

    /*
    initialData isn't currently used due to a bug. All plugin state start as {}.
    Code your components to handle an empty object, or fix the combine function in the main store.js
    */
    this._initialData = {};

    /*
    Holds all the localization messages for plugins, when they are enabled.
    */
    this._messages = {"en-US": {plugins: {}}, "fr-FR": {plugins: {}}};

    /* Holds all the components from plugins that can be injected
    into the core components (buttons, context menus, ...)
    Use an event dispatch to tie them to a component (navtree, ...)
    For the sake of simplicity we don't have an object for each type
    of component. plugin name and component name are concatenated together for ease of access.
    */
    this._registeredComponents = {};

    /*
    We keep an entry previously registered and remove into _unregisteredComponents.
    This is to remove it from the parent elements where it has been added (e.g.: NavTreeItems, ButtonControls, etc, ...)
    */
    this._unregisteredComponents = {};

    /*
    Holds all the plugin routes and can replace an existing route.
    Please be considerate of others and use your plugin name in your path,
    e.g.: <Route path="/number-range/pools:serverID component={PoolList}/>
    */
    this._registeredRoutes = {};
  }

  setMessages(messages) {
    for (let language in messages) {
      if (language in this._messages) {
        this._messages[language].plugins = {
          ...this._messages[language].plugins,
          ...messages[language].plugins
        };
      } else {
        this._messages[language] = {plugins: {...messages.plugins}};
      }
    }
  }

  registerComponent = (pluginName, component, injectAction) => {
    if (!pluginName || !component) {
      throw new Error(
        "You must enter a valid pluginName string as well as a reference to a component to register a component."
      );
    }
    this._registeredComponents[
      `plugin_${pluginName}_${component.PLUGIN_COMPONENT_NAME}`
    ] = {
      pluginName: pluginName,
      component: component,
      action: injectAction
    };
    try {
      // attempt to remove components from unregistered, if the plugin was enabled/disabled prior.
      delete this._unregisteredComponents[
        `plugin_${pluginName}_${component.PLUGIN_COMPONENT_NAME}`
      ];
    } catch (e) {
      // ignore if this is the first time.
    }
  };

  unregisterComponent = (pluginName, component, injectAction) => {
    if (!pluginName || !component) {
      throw new Error(
        "You must enter a valid pluginName string as well as a reference to a component to register a component."
      );
    }
    let fullPluginComponentName = `plugin_${pluginName}_${
      component.PLUGIN_COMPONENT_NAME
    }`;
    this._unregisteredComponents[fullPluginComponentName] = {
      pluginName: pluginName,
      component: component,
      action: injectAction
    };

    try {
      delete this._registeredComponents[
        `plugin_${pluginName}_${component.PLUGIN_COMPONENT_NAME}`
      ];
    } catch (e) {
      //don't bother if it doesn't exist.
    }
  };

  getUnregisteredComponents = () => {
    return {...this._unregisteredComponents};
  };

  getRegisteredComponents = () => {
    return {...this._registeredComponents};
  };

  getRegisteredComponent = fullPluginComponentName => {
    try {
      return injectIntl(
        this._registeredComponents[fullPluginComponentName].component
      );
    } catch (e) {}
  };

  registerRoutes = (pluginName, routes) => {
    this._registeredRoutes[pluginName] = routes;
  };

  unregisterRoutes = pluginName => {
    delete this._registeredRoutes[pluginName];
  };

  getRegisteredRoutes = () => {
    return {...this._registeredRoutes};
  };

  getArrayRoutes = () => {
    let routes = [];
    for (let plugin in this._registeredRoutes) {
      routes = routes.concat(this._registeredRoutes[plugin]);
    }
    return routes;
  };

  getMessages() {
    return {...this._messages};
  }

  setInitialData(name, dataObject) {
    this._initialData[name] = dataObject;
  }

  getInitialData() {
    return {...this._initialData};
  }

  getReducers() {
    return {...this._reducers};
  }

  registerReducer(pluginName, name, reducer, initialData) {
    this.setInitialData(name, initialData);
    this._reducers = {...this._reducers, [name]: reducer};
    if (this._emitChange) {
      this._emitChange(this.getReducers());
    }
  }

  setChangeListener(listener) {
    this._emitChange = listener;
  }
}

// export only a singleton. Entire app and all plugins share one PluginRegistry object.
export const pluginRegistry = new PluginRegistry();
