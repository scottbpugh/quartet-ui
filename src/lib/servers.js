// Copyright (c) 2018 Serial Lab
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
import Swagger from "swagger-client";
import {pluginRegistry} from "plugins/pluginRegistration";
import {store} from "store";
import actions from "actions/serversettings";

/**
 * Server - Holds data about a server and its settings, API client, ...
 */
export class Server {
  constructor(serverSettings) {
    /*
    Following object should be passed.
    {
      serverSettingName,
      hostname,
      port,
      ssl,
      serverID,
      path,
      username,
      password
    }
    */
    this.setServerData(serverSettings);
    // make saved server object available to core and plugins.
    pluginRegistry.registerServer(this);
  }

  setServerData = serverSettings => {
    /*
    Following object should be passed.
    {
      serverSettingName,
      hostname,
      port,
      ssl,
      serverID,
      path,
      username,
      password
    */
    // allow for partial updates.
    for (let key in serverSettings) {
      if (serverSettings[key] !== "" || serverSettings[key] !== undefined) {
        this[key] = serverSettings[key];
      }
    }
    this.protocol = this.ssl === true ? "https" : "http";
    this.port = !this.port ? 80 : this.port;
    this.path = !this.path && !serverSettings.path ? "" : this.path;
    this.url = this.getServerURL();
    this._client = null;
    this.manifest = null;
    this.appList = [];
  };
  getFormStructure = () => {
    return Server.getFormStructure(this.toJSON());
  };
  static getFormStructure = (initialValues = {}) => [
    {
      name: "serverSettingName",
      defaultValue: initialValues.serverSettingName || null,
      description: {
        type: "text",
        required: true,
        read_only: false,
        label: "Server Setting Name",
        help_text:
          "The label that will be used for this server connection setting."
      }
    },
    {
      name: "serverID",
      defaultValue: initialValues.serverID || null,
      description: {
        type: "hidden",
        required: false,
        read_only: false,
        label: "Server ID",
        help_text: "Hidden Server ID"
      }
    },
    {
      name: "hostname",
      defaultValue: initialValues.hostname || null,
      description: {
        type: "text",
        required: true,
        read_only: false,
        label: "Server Hostname",
        help_text:
          "A hostname or IP address, example localhost, serial-box.com, or 192.168.5.10."
      }
    },
    {
      name: "port",
      defaultValue: initialValues.port || null,
      description: {
        type: "number",
        required: true,
        read_only: false,
        label: "Port Number",
        min_value: 1,
        max_value: 65535,
        help_text: "A port to connect to. Example, 80, 8080, 443, ..."
      }
    },
    {
      name: "path",
      defaultValue: initialValues.path || "",
      description: {
        type: "text",
        required: false,
        read_only: false,
        label: "Root Path",
        helperText: "A path to interact with API (Optional), example /api"
      }
    },
    {
      name: "ssl",
      defaultValue: initialValues.ssl || false,
      description: {
        type: "boolean",
        required: false,
        read_only: false,
        label: "SSL/TLS",
        helperText: "SSL/TLS encryption"
      }
    },
    {
      name: "username",
      defaultValue: initialValues.username || null,
      description: {
        type: "text",
        required: false,
        read_only: false,
        label: "Username",
        help_text: "Basic Auth Username"
      }
    },
    {
      name: "password",
      defaultValue: initialValues.password || null,
      description: {
        type: "password",
        required: false,
        read_only: false,
        label: "Password",
        help_text: "Basic Auth Password"
      }
    }
  ];
  getArrayFields = () => {
    return [
      {name: "protocol", value: this.protocol, editable: true},
      {name: "port", value: this.port, editable: true},
      {name: "path", value: this.path, editable: true},
      {name: "ssl", value: this.ssl, editable: true},
      {name: "url", value: this.url, editable: false},
      {name: "hostname", value: this.hostname, editable: true},
      {
        name: "serverSettingsName",
        value: this.serverSettingName,
        editable: true
      },
      {name: "username", value: this.username},
      {name: "serverID", value: this.serverID}
    ];
  };

  toJSON() {
    return {
      serverID: this.serverID,
      protocol: this.protocol,
      port: this.port,
      path: this.path,
      ssl: this.ssl,
      hostname: this.hostname,
      serverSettingName: this.serverSettingName,
      url: this.url,
      appList: this.appList,
      username: this.username,
      password: this.password
    };
  }

  getServerURL = () => {
    return `${this.protocol}://${this.hostname}:${this.port}/${this.path}`;
  };

  parseSchema = () => {
    return new Promise((resolve, reject) => {
      Swagger(this.url + "schema/")
        .then(client => {
          // swagger-js client is available.
          // client.spec / client.originalSpec / client.errors
          resolve(client);
        })
        .catch(error => {
          reject(error);
        });
    });
  };

  listApps = () => {
    if (this.appList.length === 0) {
      this.getClient()
        .then(client => {
          this.appList = Object.keys(client.apis);
          // let redux know we got our data
          store.dispatch({
            type: actions.appsListUpdated,
            payload: this.toJSON()
          });
        })
        .catch(error => {
          // handle legacy.
          this.appList = ["serialbox"];
          // let redux know we got our data
          store.dispatch({
            type: actions.appsListUpdated,
            payload: this.toJSON()
          });
        });
    }
  };

  getClient = () => {
    return new Promise((resolve, reject) => {
      if (this._client) {
        resolve(this._client);
      } else {
        // we don't have a client yet.
        // Fetch it.
        this.parseSchema()
          .then(client => {
            this._client = client;
            resolve(this._client);
          })
          .catch(error => {
            reject(error);
          });
      }
    });
  };

  getManifest = () => {
    new Promise((resolve, reject) => {
      if (!this.manifest) {
        this.getClient().then(client => {
          client.apis.manifest.quartet_manifest_list().then(result => {
            this.manifest = result.body;
            resolve(result.body);
          });
        });
      } else {
        resolve(this.manifest);
      }
    });
  };
}
