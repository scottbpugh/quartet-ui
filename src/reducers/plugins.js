// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
import {handleActions} from "redux-actions";
import actions from "actions/plugins";
import {showMessage} from "lib/message";

export const initialData = () => {
  return {
    navTreeItems: [],
    plugins: {}
  };
};

export const setEnablePlugin = pluginEntries => {
  return dispatch => {
    for (let plugin in pluginEntries) {
      pluginEntries[plugin].enabled = true;
    }
    showMessage({type: "success", msg: `Plugin enabled.`});
    return dispatch({type: actions.pluginEnabled, payload: pluginEntries});
  };
};
export const setDisablePlugin = pluginEntries => {
  return dispatch => {
    for (let plugin in pluginEntries) {
      pluginEntries[plugin].enabled = false;
    }
    showMessage({type: "success", msg: "Plugin disabled."});
    return dispatch({type: actions.pluginDisabled, payload: pluginEntries});
  };
};
export default handleActions(
  {
    [actions.addToTreeServers]: (state, action) => {
      return {
        ...state,
        navTreeItems: {
          ...state.navTreeItems,
          [action.payload.pluginComponentName]: {...action.payload}
        }
      };
    },
    [actions.removeFromTreeServers]: (state, action) => {
      let navTreeItems = {...state.navTreeItems};
      delete navTreeItems[action.payload.pluginComponentName];
      return {
        ...state,
        navTreeItems: navTreeItems
      };
    },
    [actions.pluginEnabled]: (state, action) => {
      return {
        ...state,
        plugins: {...state.plugins, ...action.payload}
      };
    },
    [actions.pluginDisabled]: (state, action) => {
      return {
        ...state,
        plugins: {...state.plugins, ...action.payload}
      };
    }
  },
  {}
);
