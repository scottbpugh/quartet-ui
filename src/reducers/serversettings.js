// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import {handleActions} from "redux-actions";
import actions from "actions/serversettings";
import {showMessage} from "lib/message";
import uuidv4 from "uuid/v4";
import {pluginRegistry} from "plugins/pluginRegistration";
import {Server} from "lib/servers";

/**
 * initialData - Returns the initial data for
 * the Redux store for this form.
 * Used in index.js to initialized the store.
 *
 * It is a JSON-serializable object representing the form
 */
export const initialData = () => {
  let _initialData = {
    servers: {}, // uuid key to objects
    currentServer: null
  };
  return _initialData;
};

export const saveServer = postData => {
  return dispatch => {
    showMessage({type: "success", msg: "Your server settings were saved."});
    let server = {...postData};
    server.plugins = [];
    server.plugins.push("number-range");
    if (!postData.serverID) {
      postData.serverID = uuidv4();
    }
    dispatch({type: actions.saveServerSettings, payload: postData});
    const newServer = new Server(postData);
    pluginRegistry.registerServer(newServer);
    return dispatch({
      type: actions.serverUpdated,
      payload: JSON.stringify(newServer)
    });
  };
};

export const loadCurrentServer = serverData => {
  return dispatch => {
    return dispatch({type: actions.loadCurrentServer, payload: serverData});
  };
};

export default handleActions(
  {
    [actions.loadCurrentServer]: (state, action) => {
      return {
        ...state,
        formData: initialData(action.payload).formData
      };
    },
    [actions.saveServerSettings]: (state, action) => {
      return {
        ...state,
        servers: {...state.servers, [action.payload.serverID]: action.payload},
        formData: initialData().formData
      };
    },
    [actions.appsListUpdated]: (state, action) => {
      return {
        ...state,
        servers: {...state.servers, [action.payload.serverID]: action.payload}
      };
    }
  },
  {}
);
