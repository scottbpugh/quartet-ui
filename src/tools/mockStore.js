// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import React from "react";

import {IntlProvider} from "react-intl";
import {addLocaleData} from "react-intl";
import messages from "messages";
import en from "react-intl/locale-data/en";
import fr from "react-intl/locale-data/fr";
import {flattenMessages} from "lib/flattenMessages";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

addLocaleData([...en, ...fr]);
let defaultLocale = "en-US";
const middlewares = [thunk];

export const initialState = {
  dashboard: {notifications: []},
  serversettings: {
    servers: {
      "824ae2dd-2411-4adf-a846-d9b310a33d4b": {
        serverSettingName: "Server #1",
        serverName: "localhost",
        port: "80",
        ssl: false,
        username: "root",
        password: "toor",
        serverID: "824ae2dd-2411-4adf-a846-d9b310a33d4b"
      }
    }
  },
  intl: {
    defaultLocale: defaultLocale,
    locale: defaultLocale,
    messages: flattenMessages(messages[defaultLocale])
  },
  layout: {pageTitle: {id: "app.nav.servers"}},
  plugins: {plugins: {}, navTreeItems: []}
};

export const mockStore = configureStore(middlewares);

export const TestWrapper = ({locale, messages, children}) => {
  if (locale === undefined) {
    locale = defaultLocale;
  }
  if (messages === undefined) {
    messages = initialState.intl.messages;
  }
  return (
    <IntlProvider locale={locale} messages={messages}>
      {children}
    </IntlProvider>
  );
};

