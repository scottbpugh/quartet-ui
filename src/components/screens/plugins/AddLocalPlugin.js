// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import React, {Component} from "react";
import {Field, reduxForm} from "redux-form";
import {FormGroup, Intent, Button} from "@blueprintjs/core";
import {RightPanel} from "components/layouts/Panels";
import {FormattedMessage} from "react-intl";
import {connect} from "react-redux";
import {addLocalPlugin} from "reducers/plugins";

const FORM_NAME = "addPlugin";

class _AddLocalPlugin extends Component {
  submit = postValues => {
    // add the flag it's a local plugin.
    postValues["local"] = true;
    postValues["core"] = false;
    this.props.addLocalPlugin(postValues);
    this.props.history.push("/plugins");
  };
  render() {
    return (
      <RightPanel title={<FormattedMessage id="app.plugins.addPluginForm" />}>
        <form onSubmit={this.props.handleSubmit(this.submit.bind(this))}>
          <div>
            <FormGroup
              helperText="The name of the plugin, for programmatic use. Alphanumeric with dashes and underscores."
              label="Plugin Name"
              require={true}>
              <Field
                className="bp3-input"
                name="pluginName"
                component="input"
                type="text"
                placeholder="Plugin Name"
              />
            </FormGroup>
            <FormGroup
              helperText="The path to initialize the plugin. The file containing enable and disable."
              label="Init Path"
              require={true}>
              <Field
                className="bp3-input"
                name="initPath"
                component="input"
                type="text"
                placeholder="Init Path"
              />
            </FormGroup>
            <FormGroup
              helperText="The path to the preview thumbnail"
              label="preview">
              <Field
                className="bp3-input"
                name="preview"
                component="input"
                type="text"
                placeholder="Preview Path"
              />
            </FormGroup>
            <FormGroup
              helperText="The name as it will appear in the plugin list"
              label="Readable Name">
              <Field
                className="bp3-input"
                name="readableName"
                component="input"
                type="text"
                placeholder="Readable Name"
                require={true}
              />
            </FormGroup>
            <FormGroup
              helperText="The package path or package name in NPM."
              label="Package Path">
              <Field
                className="bp3-input"
                name="packagePath"
                component="input"
                type="text"
                placeholder="Package Path or NPM package name"
                require={true}
              />
            </FormGroup>
            <FormGroup
              helperText="The description of what this plugin does."
              label="Description">
              <Field
                className="bp3-input"
                name="description"
                component="input"
                type="text"
                placeholder="Description"
                require={true}
              />
            </FormGroup>
          </div>
          <Button type="submit">Submit</Button>
        </form>
      </RightPanel>
    );
  }
}

export default connect(
  (state, ownProps) => {
    initialValues: ownProps.initialValues || {};
  },
  {addLocalPlugin}
)(reduxForm({form: FORM_NAME})(_AddLocalPlugin));
