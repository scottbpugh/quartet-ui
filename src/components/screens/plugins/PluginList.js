// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import React, {Component} from "react";
import {RightPanel} from "components/layouts/Panels";
import {Card, Button, Tag} from "@blueprintjs/core";
import {
  setEnablePlugin,
  setDisablePlugin,
  fetchRemotePlugins
} from "reducers/plugins";
import {connect} from "react-redux";
import "./PluginList.css";
import {updateMessages} from "reducers/locales";

class Plugin extends Component {
  constructor(props) {
    super(props);
    this.state = {runtimeVersion: this.getVersionIfInstalled()};
  }
  componentDidMount() {
    this.setState({runtimeVersion: this.getVersionIfInstalled()});
  }
  componentWillReceiveProps(nextProps) {
    this.setState({runtimeVersion: this.getVersionIfInstalled()});
  }
  handleEnable = async evt => {
    let pluginModule = await window.qu4rtet.getPluginModule(
      this.props.pluginEntry
    );
    pluginModule.enablePlugin();
    this.props.setEnablePlugin({
      [this.props.pluginName]: {...this.props.pluginEntry}
    });
    this.props.updateMessages(this.props.locale);
  };
  getVersionIfInstalled = () => {
    try {
      if (this.props.pluginEntry.packagePath) {
        return qu4rtet.require(
          `${qu4rtet.pluginPath}/${
            this.props.pluginEntry.packagePath
          }/package.json`
        ).version;
      }
    } catch (e) {
      // we don't care if it's not available.
      return null;
    }
  };
  handleDisable = async evt => {
    let pluginModule = await window.qu4rtet.getPluginModule(
      this.props.pluginEntry
    );
    pluginModule.disablePlugin();
    this.props.setDisablePlugin({
      [this.props.pluginName]: {...this.props.pluginEntry}
    });
  };
  render() {
    return (
      <Card className="pt-elevation-4">
        <h5>
          {this.props.pluginEntry.readableName}{" "}
          {this.props.pluginEntry.local ? (
            <Button
              iconName="pt-icon-edit"
              className="pt-button add-plugin-button pt-intent-primary"
              onClick={this.handleEnable.bind(this)}>
              Edit
            </Button>
          ) : null}
          {!this.props.pluginEntry || !this.props.pluginEntry.enabled ? (
            <Button
              iconName="pt-icon-add"
              className="pt-button add-plugin-button pt-intent-primary"
              onClick={this.handleEnable.bind(this)}>
              Enable
            </Button>
          ) : (
            <Button
              iconName="pt-icon-remove"
              className="pt-button add-plugin-button pt-intent-primary"
              onClick={this.handleDisable.bind(this)}>
              Disable
            </Button>
          )}
        </h5>
        <div>
          {this.state.runtimeVersion ? (
            <Tag>{`version ${this.state.runtimeVersion}`}</Tag>
          ) : null}
          <div className="pt-callout pt-intent-primary pt-callout-plugin">
            {this.props.pluginEntry.preview ? (
              <img
                className="plugin-screenshot"
                src={this.props.pluginEntry.preview}
                title={this.props.pluginEntry.readableName}
                alt={this.props.pluginEntry.readableName}
              />
            ) : null}
            <p>{this.props.pluginEntry.description}</p>
          </div>
        </div>
      </Card>
    );
  }
}

export class _PluginList extends Component {
  constructor(props) {
    super(props);
    this.state = {plugins: {}};
  }
  componentDidMount() {
    this.props.fetchRemotePlugins();
  }
  componentWillReceiveProps(nextProps) {
    this.setState({plugins: nextProps.plugins});
  }
  render() {
    return (
      <RightPanel title={<formattedMessage id="app.nav.plugins" />}>
        <div className="cards-container">
          {this.state.plugins
            ? Object.keys(this.state.plugins).map(pluginName => {
                return (
                  <Plugin
                    {...this.props}
                    key={pluginName}
                    pluginName={pluginName}
                    pluginEntry={this.state.plugins[pluginName]}
                  />
                );
              })
            : null}
        </div>
      </RightPanel>
    );
  }
}

export const PluginList = connect(
  (state, ownProps) => {
    return {
      plugins: state.plugins.plugins,
      locale: state.intl.locale
    };
  },
  {setEnablePlugin, setDisablePlugin, updateMessages, fetchRemotePlugins}
)(_PluginList);
