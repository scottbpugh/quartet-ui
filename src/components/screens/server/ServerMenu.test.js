// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import React, {Component} from "react";
import "tools/mockStore"; // mock ipcRenderer, localStorage, ...
import renderer from "react-test-renderer";
import configureStore from "redux-mock-store";
import {MemoryRouter as Router, withRouter} from "react-router-dom";
import {IntlProvider, intlReducer} from "react-intl-redux";
import {Provider} from "react-redux";
import {
  mockStore,
  TestWrapper,
  initialState,
  updateRegistryIntl
} from "tools/mockStore";
import {Server} from "lib/servers";
import {FormattedMessage} from "react-intl";
import {pluginRegistry} from "plugins/pluginRegistration";
import {initialData} from "../../../reducers/serversettings";
import {ServerMenu} from "./ServerMenu";

it("renders correctly", () => {
  const store = mockStore(initialState);
  const props = {
    match: {
      params: {
        serverID: "d0246781-67c6-474b-8ab0-29de61b6e6bb"
      }
    }
  };
  pluginRegistry.registerServer(
    new Server(
      initialState.serversettings.servers[
        "d0246781-67c6-474b-8ab0-29de61b6e6bb"
      ]
    )
  );

  updateRegistryIntl(initialState.intl.locale, initialState.intl.messages);

  const serverMenu = renderer
    .create(
      <TestWrapper>
        <Provider store={store}>
          <Router>
            <ServerMenu
              server={pluginRegistry.getServer(
                "d0246781-67c6-474b-8ab0-29de61b6e6bb"
              )}
              intl={pluginRegistry.getIntl()}
            />
          </Router>
        </Provider>
      </TestWrapper>
    )
    .toJSON();
  expect(serverMenu).toMatchSnapshot();
});
