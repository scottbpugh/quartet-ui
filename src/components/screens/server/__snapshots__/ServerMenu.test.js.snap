// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<ul
  className="pt-menu"
>
  <div
    className="pt-button-group pt-minimal context-menu-control"
  >
    <button
      className="pt-button pt-icon-trash"
      disabled={undefined}
      onClick={undefined}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      small={true}
      type="button"
    />
  </div>
  <li
    className="pt-menu-header"
  >
    <h6>
      box 1
    </h6>
  </li>
  <li
    className="pt-menu-divider"
  />
  <li
    className=""
  >
    <a
      className="pt-menu-item pt-popover-dismiss"
      href={undefined}
      onClick={undefined}
      tabIndex={0}
      target={undefined}
    >
      Register User
    </a>
  </li>
  <li
    className=""
  >
    <a
      className="pt-menu-item pt-popover-dismiss"
      href={undefined}
      onClick={undefined}
      tabIndex={0}
      target={undefined}
    >
      Verify User
    </a>
  </li>
</ul>
`;
